### PR TITLE
#57 WORKSHOP

### PR DESCRIPTION
**Pull request by ivelin980**

### CODE REVIEW

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me/NodeMap.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package jv.advanced.workshop.by.me;
+
+public class NodeMap {
+	private static final int INITIAL_SIZE = 4;
+	private NodeList buckets[];
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me/NodeMap.java:5]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/DoublyLinkedList.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+			return;
+		}
+		newNode.prev = this.tail;
+		newNode.prev.next = newNode;
+		this.tail = newNode;
+	}
+
+	public boolean removeFirst() {
+		if (this.head == null) {
+			throw new NullPointerException("The list is empty!");
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/DoublyLinkedList.java:49]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/DoublyLinkedList.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		size--;
+		return true;
+	}
+
+	public boolean isEmpty() {
+		boolean isEmpty = false;
+		if (this.head == null) {
+			isEmpty = true;
+		}
+		return isEmpty;
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/DoublyLinkedList.java:72]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Deque.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+				this.elements = grow();
+			}
+		this.elements[++this.tail] = element;
+		}
+		size++;
+		return true;
+	}
+	
+	public boolean addFirst(Integer element) {
+		if(element==null) {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Deque.java:66]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Deque.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+			if (this.head == 0) {
+				this.elements = grow();
+			}
+			this.elements[--this.head] = element;
+		}
+		this.size++;
+	}
+	
+	public boolean add(Integer element) {
+		if(element==null) {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Deque.java:49]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Deque.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	public Integer peekFirst() {
+		return peek();
+	}
+
+	public Integer peekLast() {
+		return this.elements[this.tail];
+	}
+	
+	public boolean isEmpty() {
+		if(size == 0) {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Deque.java:200]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me/NodeList.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package jv.advanced.workshop.by.me;
+
+public class NodeList {
+	private Node firstElement;
+
+	public void add(Node node) {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me/NodeList.java:6]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/DoublyLinkedList.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+			return;
+		}
+		newNode.next = this.head;
+		newNode.next.prev = newNode;
+		this.head = newNode;
+
+	}
+
+	public void addLast(int element) {
+		Node newNode = new Node(element);
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/DoublyLinkedList.java:36]: too complicated

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/DoublyLinkedList.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+		private Node prev;
+		private Node next;
+
+		public Node(int element) {
+			this.element = element;
+		}
+	}
+	
+	public int size() {
+		return this.size;
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/DoublyLinkedList.java:19]: I gave example for using documentation to method addFirst

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me/Main.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package jv.advanced.workshop.by.me;
+
+public class Main {
+
+	public static void main(String[] args) {
+		NodeMap map = new NodeMap();
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me/Main.java:6]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/Main.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package jv.advanced.workshop.by.me2;
+
+public class Main {
+
+	public static void main(String[] args) {
+		DoublyLinkedList list = new DoublyLinkedList();
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me2/Main.java:6]: fixed

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Deque.java [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+	private int tail;
+	private int size;
+
+	public Deque() {
+		this.elements = new Integer[5];
+		this.head = this.elements.length / 2;
+		this.tail = this.elements.length / 2;
+	}
+
+	public void offer(Integer element) {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Deque.java:18]: I gave an example

File: javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Main.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package jv.advanced.workshop.by.me3;
+
+public class Main {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_workshop/src/jv/advanced/workshop/by/me3/Main.java:3]: fixed

**Review by ngavrailov — APPROVED** [review]

**Pull request merged**